misc/jobserver_pool.py

```diff
@@ -0,0 +1,340 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Setup a GNU Make Jobserver jobs pool then launch a command with it.
+
+On Windows, this only supports the semaphore-based scheme.
+On Posix, this uses a fifo by default, use --pipe for pipe mode.
+
+On exit, this script verifies that all job slots were returned to
+the pool, and will print an error message if this is not the case.
+
+This is useful to catch the use of broken protocol clients.
+Use the `--no-check` flag to disable this.
+
+See --help-usage for usage examples.
+"""
+import argparse
+import os
+import platform
+import subprocess
+import sys
+import typing as T
+
+# Technical note about the MAKEFLAGS values set by this script.
+#
+# All the MAKEFLAGS values created by this script begin
+# with " -j{count} ", i.e. an initial space, the "-j" characters
+# followed by a job slot count then another space.
+#
+# The initial space is only there to mimic what GNU Make 4.3
+# does. Other pool implementations do not use one and thus
+# clients should not expect it (even GNU Make doesn't seem
+# to care when used as a jobserver client).
+#
+# The {count} value is also not available in many pool
+# implementations, but is useful to better debug multi-builds
+# using this script (i.e. to verify that the pool has the
+# expected size). Protocol clients should not depend on it
+# though.
+
+_DEFAULT_NAME = "jobserver_pool"
+_IS_WINDOWS = sys.platform in ("win32", "cygwin")
+
+if _IS_WINDOWS:
+
+    try:
+        # This requires pywin32 to be installed.
+        import pywintypes
+        import win32event
+        import win32api
+    except ModuleNotFoundError:
+        print(
+            "\nERROR: Could not import Win32 API, please install pywin32, e.g. `python -m pip install pywin32`.\n",
+            file=sys.stderr,
+        )
+        raise
+
+    # It seems impossible to import a proper mypy-compatible type definition for PyHANDLE
+    # 'from pywintypes import PyHANDLE' fails stating there is no such name.
+    # 'from pywintypes import HANDLE as PyHANDLE' fails because HANDLE is a function, not a type.
+    PyHandle: T.TypeAlias = T.Any
+
+    def create_sem(
+        sem_name: str, jobs_count: int
+    ) -> T.Tuple[PyHandle, T.Dict[str, str]]:
+        """Create and initialize Win32 semaphore."""
+        assert jobs_count > 0, f"Jobs count must be strictly positive"
+        # The win32event documentation states that the first argument to CreateSemaphore()
+        # can be None to indicate default security attributes, but mypy only wants
+        # a PySECURITY_ATTRIBUTES for some reason.
+        handle = win32event.CreateSemaphore(
+            None,  # type: ignore
+            jobs_count - 1,
+            jobs_count - 1,
```

Comment:
Illegal argument for `jobs_count == 1`, argument `lMaximumCount` must be non-zero by definition.